internal/endpoint/endpoint.go

```diff
@@ -2,59 +2,87 @@
 //
 // Copyright (C) 2018 IOTech Ltd
 // Copyright (c) 2019 Intel Corporation
+// Copyright (c) 2020 Dell Inc.
 //
 // SPDX-License-Identifier: Apache-2.0
 
 package endpoint
 
 import (
+	"context"
 	"fmt"
 	"os"
 	"sync"
 	"time"
 
-	"github.com/edgexfoundry/go-mod-core-contracts/clients/types"
+	"github.com/edgexfoundry/go-mod-core-contracts/clients/interfaces"
 	"github.com/edgexfoundry/go-mod-registry/registry"
 )
 
 type Endpoint struct {
-	RegistryClient registry.Client
-	passFirstRun   bool
-	WG             *sync.WaitGroup
+	ctx            context.Context
+	wg             *sync.WaitGroup
+	RegistryClient *registry.Client
+	serviceKey     string // The key of the service as found in the service registry (e.g. Consul)
+	path           string // The path to the service's endpoint following port number in the URL
+	interval       int    // The interval in milliseconds governing how often the client polls to keep the endpoint current
+
 }
 
-func (e Endpoint) Monitor(params types.EndpointParams) chan string {
-	ch := make(chan string, 1)
-	ch <- e.fetch(params)
+func New(
+	ctx context.Context,
+	wg *sync.WaitGroup,
+	registryClient *registry.Client,
+	serviceKey string,
+	path string,
+	interval int) *Endpoint {
+
+	return &Endpoint{
+		ctx:            ctx,
+		wg:             wg,
+		RegistryClient: registryClient,
+		serviceKey:     serviceKey,
+		path:           path,
+		interval:       interval,
+	}
+}
+
+func (e Endpoint) Monitor() chan interfaces.URLStream {
+	ch := make(chan interfaces.URLStream, 1)
+	url, err := e.buildURL()
+	if err != nil {
+		_, _ = fmt.Fprintln(os.Stdout, err.Error())
+	}
+	ch <- interfaces.URLStream(url)
+
+	e.wg.Add(1)
 	go func() {
+		defer e.wg.Done()
```

Comment:
Since this is an infinite for loop, `wg.Done()` might be never called.
This line should be called before `go func()` in my opinion.